### PR TITLE
Add isAllocator

### PR DESCRIPTION
**Pull request by nordlow**
For sample usage see

https://github.com/nordlow/phobos-next/blob/master/src/nxt/dynamic_array.d#L32

Also see discussion at https://forum.dlang.org/post/uiavxptceyxsjulkxlec@forum.dlang.org.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

File: std/experimental/allocator/common.d
Comment (nordlow) [inline std/experimental/allocator/common.d:27]: Maybe it suffices to have one allocator check here instead of 4.

**Comment by atilaneves** [comment]
Why are we adding a public symbol to Phobos without tagging me? Also, what is this for?

**Comment by atilaneves** [comment]
@thewilsonator @nordlow ?

**Comment by nordlow** [comment]
> Also, what is this for?

It is supposed to be used in template-restrictions of containers as at

https://github.com/nordlow/phobos-next/blob/master/src/nxt/dynamic_array.d#L32

**Comment by atilaneves** [comment]
And why exactly can't this trait be in the library that's using it? The only place it's used in this PR is the unittest for it. I don't think this is general enough to be included in Phobos, and this shouldn't have been merged without consulting me first.

**Comment by nordlow** [comment]
> And why exactly can't this trait be in the library that's using it? The only place it's used in this PR is the unittest for it. I don't think this is general enough to be included in Phobos, and this shouldn't have been merged without consulting me first.

I believe it belongs in Phobos just like isInputRange, ... belongs in Phobos.

I propose to use isAllocator as

```d
static assert(isAllocator!A);
```

after the definition of each non-templated allocator `A` in Phobos.

**Comment by atilaneves** [comment]
> I believe it belongs in Phobos just like isInputRange and similar belong in Phobos.

`isInputRange` is actually used in Phobos. This isn't.

> I propose to use isAllocator as

Could you please explain what you think the advantage of that would be compared to now?

**Comment by ljmf00** [comment]
> Why are we adding a public symbol to Phobos without tagging me? Also, what is this for?

I think this is really not the best approach to ask for clarifications.

**Comment by nordlow** [comment]
> Could you please explain what you think the advantage of that would be compared to now?

It acts as a static contract verification check just like with ranges.

**Comment by atilaneves** [comment]
> > Why are we adding a public symbol to Phobos without tagging me? Also, what is this for?
> 
> I think this is really not the best approach to ask for clarifications.

What would your suggestion be?

**Comment by ljmf00** [comment]
> > > Why are we adding a public symbol to Phobos without tagging me? Also, what is this for?
> > 
> > 
> > I think this is really not the best approach to ask for clarifications.
> 
> What would your suggestion be?

It's all related to governance. I would say that the necessity of tagging you on every Phobos addition is not formally a requirement, even though we know its a good rule of thumb to do so. I just think you can express that clarification in other non-authoritative/soft ways. Anyway, to avoid these situation we should have a more defined governance procedures to solve ambiguities. Maybe we should have more than X approvals on simple additions such as this one and give Y amount of time before merging.

**Comment by atilaneves** [comment]
>  would say that the necessity of tagging you on every Phobos addition is not formally a requirement

Yes, it is.

**Comment by ljmf00** [comment]
> > would say that the necessity of tagging you on every Phobos addition is not formally a requirement
> 
> Yes, it is.

We should then:
1. update the contribution guidelines docs, so people can/know to tag you. (https://wiki.dlang.org/Contributing_to_Phobos#Review_process and https://wiki.dlang.org/People#Phobos)
2. have clearer governance guidelines for these situations.

**Pull request merged**
